### PR TITLE
[DEV-5399] spec metric params validation

### PR DESCRIPTION
**Pull request by gaetano-guerriero**
This change is [Reviewable](https://reviewable.io/reviews/athenianco/athenian-api/3122)

### CODE REVIEW

File: server/athenian/api/application.py [hunk 1]
```diff
@@ -931,6 +931,11 @@ def add_api(self, specification: str, ref_resolver_store: Optional[dict] = None,
 
         self.app["route_spec"] = route_spec
 
+        # collect all schemas from the spec, so that they can be used for custom validation
```
Comment (vmarkovtsev) [inline server/athenian/api/application.py:934]: Do we expect `schemas_spec` to sometimes exist, and sometimes not?

File: server/athenian/api/controllers/goal_controller.py [hunk 2]
```diff
@@ -420,6 +425,19 @@ async def _parse_update_request(
...
+    team_goals_pseudo_instance = {
+        "metric": goal[Goal.metric.name],
+        "team_goals": [tg.to_dict() for tg in update_req.team_goals],
+    }
+    try:
+        team_goals_validator.validate(team_goals_pseudo_instance)
+    except jsonschema.ValidationError as e:
+        raise ResponseError(
+            InvalidRequestError(".team_goals", f'Invalid team goals: "{e.message}"'),
+        )
```
Comment (vmarkovtsev) [inline server/athenian/api/controllers/goal_controller.py:439]: To me, this is an indicator that we are misusing OpenAPI. These checks should happen automatically. Are we underdefining the spec?

Look:
- We first deserialize JSON to dicts
- Then validate them over OpenAPI spec
- Then make the request object, validating them further
- Then we convert a part of the request back to dicts
- Then we validate those dicts over another OpenAPI spec

This is the only such place and it is not worth the trouble. We should hardcode all the checks here without feeling shame.

**Review by vmarkovtsev — CHANGES_REQUESTED** [review]

File: server/athenian/api/controllers/goal_controller.py (hunk 2 above)
Comment (gaetano-guerriero) [reply server/athenian/api/controllers/goal_controller.py:439]: The problem here is that the spec cannot express the restriction.
The restriction is about validating a piece of the request (metric_params) with something that it's not in the GoalUpdateRequest, but in the target object on the DB.
By creating a new object with can reuse the existing schema for the GoalCreateRequest

alternatives are:
A. write some procedural code to find the metric type and then manually call the correct subschema for metric_params
B. write completely procedural code for the validation of metric_params against metric
C. decide that  the ~80 lines to do B are not worth the effort and sacrifice metric_params validation for update_goal

File: server/athenian/api/application.py (hunk 1 above)
Comment (gaetano-guerriero) [reply server/athenian/api/application.py:934]: fist added specification creates the `schemas_spec` dict in the app, following ones will update the same dict; just the same as `route_spec`

Comment (gaetano-guerriero) [reply server/athenian/api/application.py]: removed now

File: server/athenian/api/controllers/goal_controller.py (hunk 2 above)
Comment (gaetano-guerriero) [reply server/athenian/api/controllers/goal_controller.py]: implemented with procedural code, PTAL

File: server/athenian/api/controllers/goal_controller.py [hunk 3]
```diff
@@ -429,3 +436,40 @@ async def _parse_update_request(
...
+                    InvalidRequestError(".team_goals", f'Invalid team goals: "{tg}"'),
+                )
+
+
+def _check_team_goals_threshold_type(
+    metric: str,
+    update_req: GoalUpdateRequest,
+    types: type | tuple[type, ...],
+) -> None:
+    for tg in [tg for tg in update_req.team_goals if tg.metric_params]:
```
Comment (vmarkovtsev) [inline server/athenian/api/controllers/goal_controller.py:472]: too many `tg`-s :smile: 
```suggestion
    for tg in update_req.team_goals:
        if not tg.metric_params:
            continue
```

File: server/athenian/api/controllers/goal_controller.py [hunk 4]
```diff
@@ -429,3 +436,40 @@ async def _parse_update_request(
         jira_priorities=jira_priorities,
         jira_issue_types=jira_issue_types,
     )
+
+
```
Comment (vmarkovtsev) [inline server/athenian/api/controllers/goal_controller.py:440]: IMO this looks nicer unless we'll have to update this frequently. Probably, we'll move them somewhere up where the metrics live, even.

**Review by vmarkovtsev — APPROVED** [review]
I think this is better than messing with the spec. Let's see how it will feel to maintain...

**Pull request merged**